Src/Newtonsoft.Json.Tests/Benchmarks/JsonTextReaderBenchmarks.cs

```diff
@@ -25,46 +25,723 @@
 
 #if HAVE_BENCHMARKS
 
+using BenchmarkDotNet.Attributes;
+using BenchmarkDotNet.Attributes.Jobs;
+using BenchmarkDotNet.Configs;
+using BenchmarkDotNet.Jobs;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
 using System.IO;
 using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading;
 using System.Xml;
-using BenchmarkDotNet.Attributes;
-using Newtonsoft.Json.Linq;
 
 namespace Newtonsoft.Json.Tests.Benchmarks
 {
+    [MemoryDiagnoser]
+    [Config(typeof(Config))]
     public class JsonTextReaderBenchmarks
     {
-        private static readonly string FloatArrayJson;
+        private static readonly string text = System.IO.File.ReadAllText(TestFixtureBase.ResolvePath("large.json"));
 
-        static JsonTextReaderBenchmarks()
+        private static readonly string EmailProperty = StringLookupTable.Shared.Get("email");
+
+        private class Config : ManualConfig
         {
-            FloatArrayJson = new JArray(Enumerable.Range(0, 5000).Select(i => i * 1.1m)).ToString(Formatting.None);
+            public Config()
+            {
+                Add(BenchmarkDotNet.Jobs.Job.LongRun.WithGcServer(true));
+            }
         }
 
-        [Benchmark]
+        [Benchmark(Baseline = true)]
         public void ReadLargeJson()
         {
-            using (StreamReader fs = System.IO.File.OpenText(TestFixtureBase.ResolvePath("large.json")))
+            var emails = new List<string>(5000);
+            using (var fs = new StringReader(text))
             using (JsonTextReader jsonTextReader = new JsonTextReader(fs))
             {
+                jsonTextReader.DateParseHandling = DateParseHandling.None;
+                jsonTextReader.PropertyNameTable = null;
+
                 while (jsonTextReader.Read())
                 {
+                    if (jsonTextReader.TokenType == JsonToken.PropertyName && jsonTextReader.Value.Equals("email"))
+                    {
+                        if (jsonTextReader.Read() && jsonTextReader.TokenType == JsonToken.String)
+                        {
+                            emails.Add((string)jsonTextReader.Value);
+                        }
+                    }
                 }
             }
         }
 
         [Benchmark]
-        public void ReadAsDecimal()
+        public void ReadLargeJson_NameTable()
+        {
+            var emails = new List<string>(5000);
+            using (var fs = new StringReader(text))
+            using (JsonTextReader jsonTextReader = new JsonTextReader(fs))
+            {
+                jsonTextReader.DateParseHandling = DateParseHandling.None;
+                jsonTextReader.PropertyNameTable = StringLookupTable.Shared;
```

Comment:
I'm curious about the implementation you're perf testing. How does it compare to the existing `DefaultNameTable`?

Reply:
My implementation is based on open hashing with quadratic probing and same hashing algorithm. Performance is +/- 5% compare to DefaultNameTable depending on test case. 

The main difference is that strings are inserted if not exists and thread safety is handled internally.

Reply:
Looks like I unintentionally pushed benchmark changes I were using for testing. I restored benchmark to original state. But you could look into my commit for name table implementation.